File: src/main/java/make/common/collection/CollectionUtil.java
package make.common.collection;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;

public class CollectionUtil {
    public static <K, V> Map<K, V> mapValues(Set<K> keys, Function<K, V> mapper) {
        Map<K, V> map = new HashMap<>(keys.size());
        for (K key: keys) {
            map.put(key, mapper.apply(key));
        }

        return map;
    }

    public static <K, V> Map<K, V> mapKeys(Set<V> values, Function<V, K> mapper) {
        Map<K, V> map = new HashMap<>(values.size());
        for (V value: values) {
            map.put(mapper.apply(value), value);
        }

        return map;
    }
}

File: src/main/kotlin/make/common/collection/CollectionUtil.kt
package make.common.collection

inline fun <T, R> Collection<T>.map(transform: (T) -> R): MutableList<R> {
    val list = ArrayList<R>(size)
    for (element in this)
        list += transform(element)

    return list
}

fun <K, V> Map<K, V>.reverse() = map { k, v -> v to k }

inline fun <K, V, K0, V0> Map<K, V>.map(transform: (K, V) -> Pair<K0, V0>): MutableMap<K0, V0> {
    val map = HashMap<K0, V0>(size)
    for (entry in this) transform(entry.key, entry.value).apply {
        map[first] = second
    }

    return map
}

inline fun <K, V, T> Collection<T>.map(transform: (T) -> Pair<K, V>): MutableMap<K, V> {
    val map = HashMap<K, V>(size)

    for (element in this)
        transform(element).apply { map[first] = second }

    return map
}

inline fun <K, V> Collection<K>.mapValues(transform: (K) -> V) =
    map { it to transform(it) }

inline fun <K, V> Collection<V>.mapKeys(transform: (V) -> K) =
    map { transform(it) to it }